nni/contrib/compression/utils/evaluator.py

```diff
@@ -1104,3 +1148,363 @@ def evaluate(self) -> Tuple[float | None, Dict[str, Any]]:
+        assert self.deepspeed_engine is not None
+        if self._ori_optimizer_step is not None:
+            self.deepspeed_engine.step = self._ori_optimizer_step
+
+    def patch_engine_step(self, before_step_tasks: List[Callable], after_step_tasks: List[Callable]):
+        assert self.deepspeed_engine is not None
+        old_step = self.deepspeed_engine.step
+
+        def patched_step(_, *args, **kwargs):
+            for task in before_step_tasks:
+                task()
+            # call origin optimizer step method
+            output = old_step(*args, **kwargs)
+            for task in after_step_tasks:
+                task()
+            return output
+
+        self.deepspeed_engine.step = types.MethodType(patched_step, self.deepspeed_engine)
+
+    def train(self, max_steps: int | None = None, max_epochs: int | None = None):
+        # deepspeed init
+        deepspeed_engine, optimizer, lr_scheduler = self.deepspeed_init(inference=False)
+        self.deepspeed_engine = deepspeed_engine
+        self.optimizer = optimizer
+        self.lr_scheduler = lr_scheduler
+        self.model = deepspeed_engine.module
+
+        assert self.deepspeed_engine is not None
+        assert self.optimizer is not None
+        assert self._training_step is not None
+
+        if hasattr(self, 'is_patch_optim_for_ds') and self.is_patch_optim_for_ds:
+            self.patch_engine_step(self.before_step_tasks, self.after_step_tasks)
+
+        self.training_func(self.deepspeed_engine, self.optimizer, self._training_step, self.lr_scheduler, max_steps, max_epochs)
+
+    def finetune(self):
+        self.train()
+
+    def evaluate(self) -> float | None | Tuple[float, Dict[str, Any]] | Tuple[None, Dict[str, Any]]:
+        # assert self.model is not None
+        if self.evaluating_func is None:
+            warn_msg = f'Did not pass evaluation_func to {self.__class__.__name__}, will return None for calling evaluate()'
+            _logger.warning(warn_msg)
+            return None
+
+        if self.deepspeed_engine is None:
+            deepspeed_engine, _, _ = self.deepspeed_init(inference=True)
+            self.deepspeed_engine = deepspeed_engine
+            self.model = self.deepspeed_engine.module
+
+        assert self.deepspeed_engine is not None
+
+        metric = self.evaluating_func(self.deepspeed_engine)
+        if isinstance(metric, dict):
+            nni_used_metric = metric.get('default', None)
+            if nni_used_metric is None:
+                warn_msg = f'Evaluation function returns a dict metric without key `default`,' + \
+                           'will return None as the model evaluation metric value.'
+                _logger.warning(warn_msg)
+            return nni_used_metric, metric
+        else:
+            return metric
+
+    def get_dummy_input(self) -> Any:
+        return self.dummy_input
+
+    def save_checkpoint(self, save_dir, tag=None, client_state={}, save_latest=True):
```

Comment:
Are these save/load functions necessary? If they are, please following the nni docstring format, so that the doc engine could render them.